[Song Ivan] iP

songivan00 opened the pull request:
# **Duke Bot**
> “If You Fail to Plan, You Are Planning to Fail” — Benjamin Franklin. [source](https://www.goodreads.com/quotes/460142-if-you-fail-to-plan-you-are-planning-to-fail#:~:text=Quote%20by%20Benjamin%20Franklin%3A%20%E2%80%9CIf,you%20are%20planning%20to%20fail!%E2%80%9D)

### Duke is an interactive _to-do list_ that is:
- Text-based
- Easy to use
- ~Fast~SUPER FAST

### A few commands you need to know:
1. _list_: see your to do list
2. _mark **task index**_: to mark the task as completed
3. _todo **task**_: to add a task to the to do list
4. _deadline **task** /by **the timing of the deadline**_: to add a task with a deadline to the to do list
5. _event **task** /at **the timing of the event**_: to add a task with a event timing
6. _delete **task index**_: to delete the task 
7. _bye_: to close the bot

### With this bot:
- [x] No more procrastination 🥳
- [x] No more late submissions 🥳
- [x] No more forgetting events 🥳

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```ruby  
public static void main(String[] args) {
    new Duke("./src/main/java/duke.txt").run();
}
```

Commit:
Add Gradle support

songivan00 committed:
Level-1

songivan00 committed:
Level-2

songivan00 committed:
Level-3

songivan00 committed:
Level-4

songivan00 committed:
A-TextUiTesting

songivan00 committed:
Level-5

songivan00 committed:
Level-6

songivan00 committed:
Adding documentation

songivan00 committed:
Level-7

songivan00 committed:
Level-8

songivan00 committed:
Merge branch 'branch-Level-7'

songivan00 committed:
Merged conflict

songivan00 committed:
More OOP

songivan00 committed:
Adding classes to Duke package

songivan00 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

songivan00 committed:
Added documentation and gradle

songivan00 committed:
Add JUnit testing

songivan00 committed:
Finishing JavaDocs

songivan00 committed:
Fix coding style

songivan00 committed:
Adding find feature

songivan00 committed:
fixed merge conflict

songivan00 committed:
Merge branch 'branch-Level-9'

Guanzhou03 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Event object at field which indicates the timings
+     */
+    protected String at;
```
I'm curious as to whether the access modifier can be private or what goes into the decision to make it protected

Guanzhou03 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,240 @@
+     */
+    public static boolean parse(String first, TaskList lst, Storage store) {
+        try {
+            if (first.equals("bye")) {
+                Ui.byeMessage();
+                return true;
+            } else if (first.equals("list")) {
+                listMessage(lst);
+                return false;
+            } else if (first.length() == 6 && first.substring(0, 4).equals("mark")) {
+                markMessage(first, lst, store);
+                return false;
+            } else if (first.length() == 8 && first.substring(0, 6).equals("unmark")) {
+                unmarkMessage(first, lst, store);
+                return false;
+            } else if (first.length() >= 4 && first.substring(0, 4).equals("todo")) {
+                todoMessage(first, lst, store);
+                return false;
+            } else if (first.length() >= 8 && first.substring(0, 8).equals("deadline")) {
+                deadlineMessage(first, lst, store);
+                return false;
+            } else if (first.length() >= 5 && first.substring(0, 5).equals("event")) {
+                eventMessage(first, lst, store);
+                return false;
+            } else if (first.length() == 8 && first.substring(0, 6).equals("delete")) {
+                deleteMessage(first, lst, store);
+                return false;
+            } else if (first.length() >= 4 && first.substring(0, 4).equals("find")) {
+                findMessage(first, lst, store);
+                return false;
+            }
```
LGTM, can consider using switch case to reduce the number of lines

songivan00 replied:
Thanks for the suggestions will implement that.

Guanzhou03 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,240 @@
+    public static void markMessage(String str, TaskList lst, Storage storage) throws DukeException, IOException {
+        char index = str.charAt(5);
+        int number = Integer.parseInt(String.valueOf(index));
+        if(number > lst.count()) {
```
Good to have a space after the "if"

Guanzhou03 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,240 @@
+        }
+    }
+
+    /**
+     * Finds a task based on a keyword
+     *
+     * @param str The users input
+     * @param lst The tasklist
+     * @param storage the storage object
+     */
```
Good that all public methods have JavaDoc comments

songivan00 replied:
Thanks!

onepersonhere commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Deadline object by field which indicates the deadline
+     */
+    protected LocalDate by;
```
Perhaps a name should explain the named entity to the reader accurately and at a sufficient level of detail?
```suggestion
    protected LocalDate byDate;
```

songivan00 replied:
Youre right, thank you!

onepersonhere commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,55 @@
+package duke;
+
+import java.io.FileNotFoundException;
+import java.util.*;
```
Perhaps imported classes should always be listed explicitly?

onepersonhere commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,240 @@
+     * @param lst The tasklist
+     * @param storage the storage object
+     */
+    public static void todoMessage(String str, TaskList lst, Storage storage) throws DukeException, IOException {
```
Perhaps use verb for actions?
```suggestion
    public static void parseTodoMessage(String str, TaskList lst, Storage storage) throws DukeException, IOException {
```

onepersonhere replied:
I found that many of the method names have the same issue.

songivan00 replied:
thanks for pointing that out! will change them :)

onepersonhere commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,103 @@
+            String currentLine = sc.nextLine();
+            String[] stringDetails = currentLine.split("\\|");
+            String taskType = stringDetails[0];
+            String isMark = stringDetails[1];
```
Perhaps name it in such a way to not sound like a boolean variable?

songivan00 replied:
ok! thank you!

onepersonhere commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,105 @@
+     *
+     * @param tsk The task
+     */
+    public void add(Task tsk) {
```
Perhaps name should explain the named entity to the reader accurately and at a sufficient level of detail?
```suggestion
    public void addTask(Task tsk) {
```

onepersonhere replied:
I found many similar methods with names of insufficient details.

onepersonhere reviewed (commented):
Overall, I found your code easy to read for the most part except for the huge else-if chunk. 
I noted a few minor coding standard & naming violations too. 
Consider renaming some methods/variables.

songivan00 committed:
GUI for Duke

songivan00 committed:
Fixed OOP and coding style

songivan00 committed:
Added assertions

Added assertions to getUserDialog and getDukeDialog to ensure that no empty String or Image is passed into these methods

Previously, such method could cause error if no text or pictures were passed.

Assertions can help define assumptions about the program states soe that the runtime can verify them.

songivan00 committed:
Improve Code Quality

Improve the quality of my code, mainly changing methods with more than 30 LOC and ensuring arrowheads are minimized.
Worked on the JavaDocs as well

An improved code quality makes the code a lot easier to read and debug when an error occurs.

songivan00 committed:
Merge pull request #2 from songivan00/Branch-A-Assertions

Add assertions

songivan00 committed:
Merge pull request #3 from songivan00/branch-A-CodeQuality

Improve code quality

songivan00 committed:
Add postpone feature

Add postpone feature for deadline and event types

This allows user to edit the timings of deadlines and events

songivan00 committed:
Merge pull request #4 from songivan00/branch-B-Snooze

Add postpone feature

songivan00 committed:
Update Brace Style

songivan00 committed:
Edit Readme

songivan00 committed:
Edit README

songivan00 committed:
add feature in readme
